proposals/NNNN-uniformity-qualifiers.md

```diff
@@ -0,0 +1,132 @@
+(SIMD) processor. When executing on a SIMD processor the program may execute
+where a single instruction produces results for multiple threads of execution
+from the source programming model. This is sometimes referred to as Single
+Instruction Multiple Threads (SIMT).
+
+Under HLSL's execution model, groups of threads form hierarchical scopes:
+* A _dispatch_ represents the full set of threads spawned from a CPU API
+  invocation.
+* A _thread group_ represents a subset of a dispatch that can execute
+  concurrently.
+* A _wave_ represents a subset of a thread group that represents in a single
+  SIMD processor.
+* A _quad_ represents a grouping of four adjacent threads in a wave.
+
+When a shader is executing threads concurrently on one or more processing cores,
+an emergent property of _uniformity_ exists within the thread group, wave and
+quad scopes.
+
+Uniformity can refer to data or control flow. If a variable has the same value
+across all threads in a scope, it is said to be _uniform_ across that scope.
+Similarly if all threads within a scope are actively executing instructions
+within a control flow block, the control flow is said to be _uniform control
+flow_ across that scope.
+
+## Motivation
+
+Uniformity of data and control flow are central concepts to SIMT execution
+models, and is required for correct execution of shader programs. Despite
+the importance of this fundamental property it is not represented in any
+explicit way in the HLSL language.
+
+This proposal, seeks to address that by introducing core concepts around
+uniformity to HLSL's type system and programming model.
+
+## Proposed solution
+
+### Uniformity as a Type Qualifier
+
+This proposal introduces a new set of type qualifiers to represent the different
+scopes of uniformity:
+* `group_uniform`
+* `simd_uniform`
+* `quad_uniform`
+* non-uniform (default state with no associated keyword)
+
+`group_uniform` is the highest scoping of uniformity, and implies all other
+scopes. `simd_uniform` implies `quad_uniform`.
+
+A new "UniformityReduction" cast will reduce the uniform scope allowing
+conversion of one uniformity scope to another uniformity scope as long as the
+source scope has a greater uniformity scope.
+
+Any GLValue with a uniformity scope can be implicitly converted to a GLValue
+with reduced uniform scope or no uniformity scope.
+
+Any PRValue with a uniformity scope can be implicitly converted to a PRValue
+with reduced uniform scope or no uniformity scope.
+
+No implicit or explicit cast can increase uniformity scope.
+
+HLSL library functionality that produces uniform results will be updated to
+produce appropriately qualified uniform types. These functions can produce
+uniform values from non-uniform inputs. For example:
+
+```hlsl
+simd_uniform bool WaveActiveAllTrue(bool);
+quad_uniform bool QuadAny(bool);
+```
+
+Compile-time constants and Groupshared variable declarations imply
+`group_uniform` uniformity.

```

Comment:
Should constant buffer/SRV data also be automatically marked this way? They're supposed to be read-only and non-volatile for the lifetime of the shader AFAIU.